core/src/main/java/org/mskcc/oncotree/utils/TumorTypesUtil.java

```
@@ -137,33 +136,60 @@ public static InputStream getTumorTypeInputStreamFromGitHub(Version version) {
+            TumorType tumorType = initTumorType(node, version);
+            allNodes.put(tumorType.getCode(), tumorType);
+        }
 
-        //Iterate each row and assign tumor type to parent following the order of appearing
-        for (String[] row : allRows.subList(1, allRows.size())) {
-            tumorType.setChildren(attachTumorType(tumorType.getChildren(), row, 0, version, "TISSUE"));
+        // now we have all nodes, fill in children
+        for (TumorType tumorType : allNodes.values()) {
+            // TISSUE is root and has no parent, skip
+            if (!tumorType.getCode().equals("TISSUE")) {
+                if (tumorType.getParent() == null) {
+                    logger.debug("Parent is null for tumor type code '" +
+                        tumorType.getCode() + "'.  Adding to 'TISSUE' node.");
+                    tumorType.setParent(root.getCode());
+                    root.addChild(tumorType);
+                } else if (allNodes.containsKey(tumorType.getParent())) {
+                    TumorType parent = allNodes.get(tumorType.getParent());
+                    parent.addChild(tumorType);
+                } else {
+                    logger.error("Could not find parent '" + 
+                        tumorType.getParent() + "' for tumor type code '" + 
+                        tumorType.getCode() + "'.");
+                }
+            }
         }
 
-        //Attach a root node in the JSON file
-        tumorTypes.put("TISSUE", tumorType);
+        return tree;
+    }
 
-        return tumorTypes;
+    private static TumorType initTumorType(OncoTreeNode oncoTreeNode, Version version) {
+        // we do not have level or tissue
+        TumorType tumorType = new TumorType();
+        if (oncoTreeNode.getMainType() != null) {
+            tumorType.setMainType(MainTypesUtil.getOrCreateMainType(oncoTreeNode.getMainType(), version));
```

I wonder how maintype would be "created" if it is not known. For us, I think we may not want to guess and want to throw an exception if the mainType is not defined .. or leave it as undefined. Some of the primary level nodes have a (as of yet) undefined maintype and we plan to show those with cancer_type "NA" in the portal interface